File: smells/unrealistic/x-ray-specs/x-ray-specs.js
// Subject under test
var _ = require('lodash')
function SeatMap (ticket, originalSeat) {
  this.fareClass = ticket.fareClass
  this.__currentSeat = originalSeat
  this.__approvals = {}
}

SeatMap.prototype.moveTo = function (newSeat) {
  if (!_.has(this.__approvals, this.fareClass + '.' + newSeat)) {
    this.__qualifyFareClassForSeat(newSeat)
  }

  if (this.__approvals[this.fareClass][newSeat]) {
    this.__currentSeat = newSeat
  }
}

// Test
module.exports = {
  beforeEach: function () {
    this.ticket = {fareClass: 'M'}
    this.subject = new SeatMap(this.ticket, '18D')
  },
  approveIfBehindRowTen: function () {
    this.subject.moveTo('11B')

    assert.equal(this.subject.__approvals['M']['11B'], true)
    assert.equal(this.subject.__currentSeat, '11B')
  },
  denyIfAheadOfRowTen: function () {
    this.subject.moveTo('9J')

    assert.equal(this.subject.__approvals['M']['9J'], false)
    assert.equal(this.subject.__currentSeat, '18D')
  },
  willShortCircuitApprovalProcessWhenMemoized: function () {
    this.subject.__approvals['M'] = {'Havanna': 'Sure, why not'}

    this.subject.moveTo('Havanna')

    assert.equal(this.subject.__approvals['M']['Havanna'], 'Sure, why not')
    assert.equal(this.subject.__currentSeat, 'Havanna')
  }
}

// Fake production implementations to simplify example test of subject
SeatMap.prototype.__qualifyFareClassForSeat = function (seat) {
  var allowed = parseInt(seat.match(/^(\d+)/)[0], 10) > 10
  _.set(this.__approvals, this.fareClass + '.' + seat, allowed)
}

File: smells/unrealistic/x-ray-specs/x-ray-specs.rb
# Subject under test
class MutableSeatMap
  attr_reader :approvals, :current_seat

  def initialize(ticket, original_seat)
    @fare_class = ticket.fare_class
    @current_seat = original_seat
    @approvals = {}
  end

  def move_to(new_seat)
    if !@approvals[@fare_class] || !@approvals[@fare_class][new_seat]
      qualify_fare_class_for_seat!(new_seat)
    end

    if @approvals[@fare_class][new_seat]
      @current_seat = new_seat
    end
  end
end

# Test
class XRaySpecs < SmellTest
  def setup
    @ticket = OpenStruct.new(fare_class: "M")
    @subject = MutableSeatMap.new(@ticket, "18D")
    super
  end

  def test_approve_if_behind_row_ten
    @subject.move_to("11B")

    assert_equal true, @subject.approvals["M"]["11B"]
    assert_equal "11B", @subject.current_seat, "11B"
  end

  def test_deny_if_ahead_of_row_ten
    @subject.move_to("9J")

    assert_equal false, @subject.approvals["M"]["9J"]
    assert_equal "18D", @subject.current_seat
  end

  def test_will_short_circuit_approval_process_when_memoized
    @subject.approvals["M"] = {"Havanna" => "Sure, why not"}

    @subject.move_to("Havanna")

    assert_equal "Sure, why not", @subject.approvals["M"]["Havanna"]
    assert_equal "Havanna", @subject.current_seat
  end
end

# Fake production implementations to simplify example test of subject
class MutableSeatMap
  def qualify_fare_class_for_seat!(seat)
    allowed = seat.match(/^(\d+)/)[0].to_i > 10
    @approvals[@fare_class] ||= {}
    @approvals[@fare_class][seat] = allowed
  end
end


